programs/drift/src/instructions/admin.rs

```diff
@@ -4970,6 +4978,291 @@ pub fn handle_add_amm_constituent_data<'info>(
+    let constituent_out_token_account = &ctx.accounts.constituent_out_token_account;
+
+    let mut in_constituent = ctx.accounts.in_constituent.load_mut()?;
+    let mut out_constituent = ctx.accounts.out_constituent.load_mut()?;
+
+    let remaining_accounts = &mut ctx.remaining_accounts.iter().peekable();
+    let out_token_program = get_token_interface(remaining_accounts)?;
+
+    let in_mint = get_token_mint(remaining_accounts)?;
+    let out_mint = get_token_mint(remaining_accounts)?;
+
+    validate!(
+        in_mint.is_some(),
+        ErrorCode::InvalidSwap,
+        "EndLpSwap must have a mint for in token passed in"
+    )?;
+
+    validate!(
+        out_mint.is_some(),
+        ErrorCode::InvalidSwap,
+        "EndLpSwap must have a mint for out token passed in"
+    )?;
+
+    let in_mint = in_mint.unwrap();
+    let out_mint = out_mint.unwrap();
+
+    // Validate that the passed mint is accpetable
+    let mint_key = out_mint.key();
+    validate!(
+        mint_key == constituent_out_token_account.mint,
+        ErrorCode::InvalidSwap,
+        "mint passed to EndLpSwap does not match the mint constituent out token account"
+    )?;
+
+    let mint_key = in_mint.key();
+    validate!(
+        mint_key == constituent_in_token_account.mint,
+        ErrorCode::InvalidSwap,
+        "mint passed to EndLpSwap does not match the mint constituent in token account"
+    )?;
+
+    // Residual of what wasnt swapped
+    if signer_in_token_account.amount > in_constituent.flash_loan_initial_token_amount {
+        let residual = signer_in_token_account
+            .amount
+            .safe_sub(in_constituent.flash_loan_initial_token_amount)?;
+
+        controller::token::receive(
+            &ctx.accounts.token_program,
+            signer_in_token_account,
+            constituent_in_token_account,
+            &admin_account_info,
+            residual,
+            &Some(in_mint),
+        )?;
+    }
+
+    // Whatever was swapped
+    if signer_out_token_account.amount > out_constituent.flash_loan_initial_token_amount {
+        let residual = signer_out_token_account
+            .amount
+            .safe_sub(out_constituent.flash_loan_initial_token_amount)?;
+
+        if let Some(token_interface) = out_token_program {
+            receive(
+                &token_interface,
+                signer_out_token_account,
+                constituent_out_token_account,
+                &admin_account_info,
+                residual,
+                &Some(out_mint),
+            )?;
+        } else {
+            receive(
+                &ctx.accounts.token_program,
+                signer_out_token_account,
+                constituent_out_token_account,
+                &admin_account_info,
+                residual,
+                &Some(out_mint),
+            )?;
+        }
+    }
+
+    // Update the balance on the token accounts for after swap
+    out_constituent.sync_token_balance(constituent_out_token_account.amount);
```

Comment:
do we need any sort of validation that the swap doesnt balance amm in wrong direction?